ArcGISRuntimeSDKQt_CppSamples/Routing/RouteAroundBarriers/RouteAroundBarriers.cpp

```diff
@@ -0,0 +1,307 @@
+  {
+    Point clickedPoint = m_mapView->screenToLocation(e.x(), e.y());
+    if (m_addStops)
+    {
+      // add stop to list of stops
+      Stop stopPoint(clickedPoint);
+      m_stopsList << stopPoint;
+
+      // create a marker symbol and graphics, and add the graphics to the graphics overlay
+      TextSymbol* textSymbol = new TextSymbol(QString::number(m_stopsList.size()), Qt::white, 16, HorizontalAlignment::Center, VerticalAlignment::Bottom, this);
+      textSymbol->setOffsetY(m_pinSymbol->height() / 2);
+      CompositeSymbol* newStopSymbol = new CompositeSymbol(QList<Symbol*>{m_pinSymbol, textSymbol}, this);
+
+      Graphic* stopGraphic = new Graphic(clickedPoint, newStopSymbol, this);
+      m_stopsOverlay->graphics()->append(stopGraphic);
+
+      createAndDisplayRoute();
+    }
+    else if (m_addBarriers)
+    {
+      Polygon barrierPolygon = GeometryEngine::buffer(clickedPoint, 500);
+      PolygonBarrier barrier(barrierPolygon);
+      m_barriersList << barrier;
+
+      Graphic* barrierGraphic = new Graphic(barrierPolygon, m_barrierSymbol);
+      m_barriersOverlay->graphics()->append(barrierGraphic);
+
+      createAndDisplayRoute();
+    }
+  });
+
+  connect (m_routeTask, &RouteTask::solveRouteCompleted, this, [this](QUuid, const RouteResult routeResult)
+  {
+    if (routeResult.isEmpty())
+      return;
+
+    Route route = routeResult.routes()[0];
+    Geometry routeGeometry = route.routeGeometry();
+    Graphic* routeGraphic = new Graphic(routeGeometry, this);
+    m_routeOverlay->graphics()->append(routeGraphic);
+
+    m_directions = route.directionManeuvers(this);
+    emit directionsChanged();
+  });
+}
+
+void RouteAroundBarriers::createAndDisplayRoute()
+{
+  if (m_stopsList.size() > 1)
+  {
+    // clear the previous route, if it exists
+    if (m_routeOverlay)
+      m_routeOverlay->graphics()->clear();
+
+    // clear the directions list
+    if (m_directions)
+    {
+      delete m_directions;
+      m_directions = nullptr;
+    }
+
+    m_routeParameters.setStops(m_stopsList);
+    m_routeParameters.setPolygonBarriers(m_barriersList);
+    m_routeParameters.setFindBestSequence(m_findBestSequence);
+    m_routeParameters.setPreserveFirstStop(m_preserveFirstStop);
+    m_routeParameters.setPreserveLastStop(m_preserveLastStop);
+
+    m_routeTask->solveRoute(m_routeParameters);
+  }
+}
+
+void RouteAroundBarriers::clearRouteAndGraphics()
+{
+  // clear stops from route parameters and stops list
+  m_routeParameters.clearStops();
+  m_stopsList.clear();
+
+  // clear barriers
+  m_routeParameters.clearPolygonBarriers();
+  m_barriersList.clear();
+
+  // clear directions list
+  if (m_directions)
+  {
+    delete m_directions;
+    m_directions = nullptr;
+    emit directionsChanged();
+  }
+
+  // clear graphics overlays
+  for (GraphicsOverlay* overlay : *m_mapView->graphicsOverlays())
+  {
+    if (overlay)
+      overlay->graphics()->clear();
```

Comment:
we need to be careful here - clearing the graphics overlay doesn't actually delete the graphics that they contain so this will leak. You can either iterate the overlay first and delete all graphics or store the graphics as members to delete them individually. Another technique that you can use is to create a `QObject* m_graphicsParent` and use that as a parent for any graphics you want to be able to deallocate. Then you just have to delete the parent object here.